### ISSUE TITLE
Assigning jet variations by setitem quite slow compared to bulk assignment by awkward.zip

### ISSUE THREAD

**Issue by lgray** [scikit-hep/awkward#538]
Trying to build jet systematics ran into some strange timing performance results. In particular this is assigning lists of variations of jets back into the list of jets, assignment by setattr is much more straightforward to use and better for bookkeeping, and I can imagine variations like this being written by end users for other corrections. The difference in performance is rather striking.

This is with 40 events / 188 jets, so most of the time is spent constructing descriptions of arrays.

```python3
import random
import time
import copy
import awkward1
from coffea.nanoevents import NanoEventsFactory

fname = 'https://raw.githubusercontent.com/CoffeaTeam/coffea/master/tests/samples/nano_dy.root'
factory = NanoEventsFactory.from_file(fname)
events = factory.events()

decorated_jets = awkward1.flatten(events.Jet)

names = {'%06x' % random.randrange(16**6) for i in range(50)}

print('number of variants', len(names))

tic = time.time()

syst_dict = {}
for name in names:
    syst_dict[name] = awkward1.zip({'up': awkward1.flatten(events.Jet),
                                    'down': awkward1.flatten(events.Jet)},
                                    depth_limit=1,
                                    with_name='JetSystematic')
toc = time.time()

print('time to make variants:', toc-tic)

tic = time.time()
for name in names:
    decorated_jets[name] = syst_dict[name]
toc = time.time()
print('time to assign with __setitem__:', toc-tic)


tic = time.time()
old_fields = awkward1.fields(decorated_jets)
syst_dict.update({field: decorated_jets[field] for field in old_fields})
decorated_jets = awkward1.zip(syst_dict, depth_limit=1, with_name='Jet')
toc = time.time()
print('time to assign with rezip:', toc-tic)
```

results in:
```
number of variants 50
time to make variants: 0.40229296684265137
time to assign with __setitem__: 8.58899211883545
time to assign with rezip: 0.2941272258758545
```

Not to mention the latter does some very funky thing to the type of the array.
On a not-too-old macbook pro.

**Comment by lgray** [comment]
Here's a profile for the `__setitem__` version:
```

  _     ._   __/__   _ _  _  _ _/_   Recorded: 15:45:07  Samples:  10278
 /_//_/// /_\ / //_// / //_'/ //     Duration: 10.409    CPU time: 10.359
/   _/                      v3.2.0

Program: example_for_jim.py

10.409 <module>  example_for_jim.py:1
└─ 10.409 __setitem__  awkward1/highlevel.py:966
   └─ 10.346 with_field  awkward1/operations/structure.py:479
      ├─ 8.330 broadcast_and_apply  awkward1/_util.py:492
      │  ├─ 7.106 apply  awkward1/_util.py:549
      │  │  ├─ 2.904 apply  awkward1/_util.py:549
      │  │  │  ├─ 1.463 of  awkward1/nplike.py:10
      │  │  │  │  └─ 1.463 kernels  awkward1/operations/convert.py:533
      │  │  │  │     └─ 1.461 recursive_walk  awkward1/_util.py:1287
      │  │  │  │        └─ 1.453 recursive_walk  awkward1/_util.py:1287
      │  │  │  │           └─ 1.414 recursive_walk  awkward1/_util.py:1287
      │  │  │  │              ├─ 1.228 recursive_walk  awkward1/_util.py:1287
      │  │  │  │              │  ├─ 0.631 apply  awkward1/operations/convert.py:564
      │  │  │  │              │  │  ├─ 0.366 [self]  
      │  │  │  │              │  │  └─ 0.261 isinstance  <built-in>:0
      │  │  │  │              │  ├─ 0.317 isinstance  <built-in>:0
      │  │  │  │              │  └─ 0.280 [self]  
      │  │  │  │              └─ 0.119 [self]  
      │  │  │  └─ 1.438 getfunction  awkward1/operations/structure.py:541
      │  │  │     └─ 1.438 of  awkward1/nplike.py:10
      │  │  │        └─ 1.438 kernels  awkward1/operations/convert.py:533
      │  │  │           └─ 1.437 recursive_walk  awkward1/_util.py:1287
      │  │  │              └─ 1.430 recursive_walk  awkward1/_util.py:1287
      │  │  │                 └─ 1.387 recursive_walk  awkward1/_util.py:1287
      │  │  │                    └─ 1.223 recursive_walk  awkward1/_util.py:1287
      │  │  │                       ├─ 0.630 apply  awkward1/operations/convert.py:564
      │  │  │                       │  ├─ 0.338 [self]  
      │  │  │                       │  └─ 0.279 isinstance  <built-in>:0
      │  │  │                       ├─ 0.312 isinstance  <built-in>:0
      │  │  │                       └─ 0.281 [self]  
      │  │  ├─ 1.488 getfunction  awkward1/operations/structure.py:541
      │  │  │  └─ 1.488 of  awkward1/nplike.py:10
      │  │  │     └─ 1.487 kernels  awkward1/operations/convert.py:533
      │  │  │        └─ 1.483 recursive_walk  awkward1/_util.py:1287
      │  │  │           └─ 1.483 recursive_walk  awkward1/_util.py:1287
      │  │  │              └─ 1.479 recursive_walk  awkward1/_util.py:1287
      │  │  │                 └─ 1.436 recursive_walk  awkward1/_util.py:1287
      │  │  │                    ├─ 1.269 recursive_walk  awkward1/_util.py:1287
      │  │  │                    │  ├─ 0.619 apply  awkward1/operations/convert.py:564
      │  │  │                    │  │  ├─ 0.352 [self]  
      │  │  │                    │  │  └─ 0.255 isinstance  <built-in>:0
      │  │  │                    │  ├─ 0.345 [self]  
      │  │  │                    │  └─ 0.305 isinstance  <built-in>:0
      │  │  │                    └─ 0.115 [self]  
      │  │  ├─ 1.432 of  awkward1/nplike.py:10
      │  │  │  └─ 1.432 kernels  awkward1/operations/convert.py:533
      │  │  │     └─ 1.430 recursive_walk  awkward1/_util.py:1287
      │  │  │        └─ 1.427 recursive_walk  awkward1/_util.py:1287
      │  │  │           └─ 1.422 recursive_walk  awkward1/_util.py:1287
      │  │  │              └─ 1.371 recursive_walk  awkward1/_util.py:1287
      │  │  │                 ├─ 1.191 recursive_walk  awkward1/_util.py:1287
      │  │  │                 │  ├─ 0.600 apply  awkward1/operations/convert.py:564
      │  │  │                 │  │  ├─ 0.338 [self]  
      │  │  │                 │  │  └─ 0.245 isinstance  <built-in>:0
      │  │  │                 │  ├─ 0.317 isinstance  <built-in>:0
      │  │  │                 │  └─ 0.274 [self]  
      │  │  │                 └─ 0.132 [self]  
      │  │  └─ 1.167 [self]  
      │  └─ 1.216 <genexpr>  awkward1/_util.py:998
      │     └─ 1.216 broadcast_unpack  awkward1/_util.py:1025
      │        └─ 1.175 [self]  
      ├─ 1.018 wrap  awkward1/_util.py:393
      │  └─ 1.017 __init__  awkward1/highlevel.py:209
      │     └─ 0.957 find_caches  awkward1/_util.py:1346
      │        └─ 0.957 recursive_walk  awkward1/_util.py:1287
      │           └─ 0.949 recursive_walk  awkward1/_util.py:1287
      │              └─ 0.921 recursive_walk  awkward1/_util.py:1287
      │                 └─ 0.816 recursive_walk  awkward1/_util.py:1287
      │                    ├─ 0.303 isinstance  <built-in>:0
      │                    ├─ 0.294 [self]  
      │                    └─ 0.219 apply  awkward1/_util.py:1349
      │                       └─ 0.204 [self]  
      └─ 0.963 [self]  
```

as well as the awkward.zip version:
```

  _     ._   __/__   _ _  _  _ _/_   Recorded: 15:45:18  Samples:  348
 /_//_/// /_\ / //_// / //_'/ //     Duration: 0.356     CPU time: 0.354
/   _/                      v3.2.0

Program: example_for_jim.py

0.356 <module>  example_for_jim.py:1
├─ 0.267 zip  awkward1/operations/structure.py:263
│  ├─ 0.226 broadcast_and_apply  awkward1/_util.py:492
│  │  ├─ 0.175 apply  awkward1/_util.py:549
│  │  │  ├─ 0.064 apply  awkward1/_util.py:549
│  │  │  │  ├─ 0.057 of  awkward1/nplike.py:10
│  │  │  │  │  └─ 0.057 kernels  awkward1/operations/convert.py:533
│  │  │  │  │     └─ 0.057 recursive_walk  awkward1/_util.py:1287
│  │  │  │  │        └─ 0.055 recursive_walk  awkward1/_util.py:1287
│  │  │  │  │           ├─ 0.046 recursive_walk  awkward1/_util.py:1287
│  │  │  │  │           │  ├─ 0.025 apply  awkward1/operations/convert.py:564
│  │  │  │  │           │  │  ├─ 0.013 isinstance  <built-in>:0
│  │  │  │  │           │  │  └─ 0.012 [self]  
│  │  │  │  │           │  ├─ 0.012 isinstance  <built-in>:0
│  │  │  │  │           │  └─ 0.009 [self]  
│  │  │  │  │           └─ 0.007 [self]  
│  │  │  │  └─ 0.006 <lambda>  awkward1/operations/structure.py:392
│  │  │  ├─ 0.062 of  awkward1/nplike.py:10
│  │  │  │  └─ 0.062 kernels  awkward1/operations/convert.py:533
│  │  │  │     └─ 0.061 recursive_walk  awkward1/_util.py:1287
│  │  │  │        └─ 0.059 recursive_walk  awkward1/_util.py:1287
│  │  │  │           └─ 0.058 recursive_walk  awkward1/_util.py:1287
│  │  │  │              └─ 0.053 recursive_walk  awkward1/_util.py:1287
│  │  │  │                 ├─ 0.025 apply  awkward1/operations/convert.py:564
│  │  │  │                 │  ├─ 0.014 [self]  
│  │  │  │                 │  └─ 0.011 isinstance  <built-in>:0
│  │  │  │                 ├─ 0.017 isinstance  <built-in>:0
│  │  │  │                 └─ 0.011 [self]  
│  │  │  └─ 0.044 [self]  
│  │  ├─ 0.045 <genexpr>  awkward1/_util.py:998
│  │  │  └─ 0.045 broadcast_unpack  awkward1/_util.py:1025
│  │  │     └─ 0.044 [self]  
│  │  └─ 0.006 broadcast_pack  awkward1/_util.py:1001
│  └─ 0.039 wrap  awkward1/_util.py:393
│     └─ 0.039 __init__  awkward1/highlevel.py:209
│        └─ 0.037 find_caches  awkward1/_util.py:1346
│           └─ 0.037 recursive_walk  awkward1/_util.py:1287
│              └─ 0.037 recursive_walk  awkward1/_util.py:1287
│                 └─ 0.037 recursive_walk  awkward1/_util.py:1287
│                    └─ 0.035 recursive_walk  awkward1/_util.py:1287
│                       ├─ 0.017 [self]  
│                       ├─ 0.009 isinstance  <built-in>:0
│                       └─ 0.009 apply  awkward1/_util.py:1349
│                          └─ 0.008 [self]  
└─ 0.085 <dictcomp>  example_for_jim.py:48
   └─ 0.085 __getitem__  awkward1/highlevel.py:614
      ├─ 0.046 wrap  awkward1/_util.py:393
      │  └─ 0.046 __init__  awkward1/highlevel.py:209
      │     ├─ 0.036 find_caches  awkward1/_util.py:1346
      │     │  └─ 0.035 recursive_walk  awkward1/_util.py:1287
      │     │     └─ 0.035 recursive_walk  awkward1/_util.py:1287
      │     │        └─ 0.033 recursive_walk  awkward1/_util.py:1287
      │     │           ├─ 0.014 [self]  
      │     │           ├─ 0.010 isinstance  <built-in>:0
      │     │           └─ 0.009 apply  awkward1/_util.py:1349
      │     └─ 0.007 arrayclass  awkward1/_util.py:141
      │        └─ 0.005 [self]  
      └─ 0.039 [self]  
```

**Comment by jpivarski** [comment]
First observation: 78% of the `__setitem__` time is spent in `broadcast_and_apply` (Python) and 0.1% of the time is spent in `setitem_field` (C++, copying RecordArrays). So the immutability of RecordArrays, requiring them to be copied every time, is not at fault. The fact that it's in Python will make it easier to narrow in on.

**Comment by jpivarski** [comment]
It looks like about 1/4 of the time is identifying whether this is a CPU array or a GPU array and 3/4 is in what ought to be trivial `getitem_at` and `getitem_slice` operations on RecordArrays. Both of these are disturbing, but the CPU vs GPU determination most of all. I'll start by looking into that.

(CPU vs GPU is to determine whether we can use NumPy or CuPy.)

**Comment by jpivarski** [comment]
Step 1, which affects 1/4 of the time but shouldn't be allowed to be a bottleneck (deciding between NumPy and CuPy _ought_ to be free): I moved it from a Python `recursively_apply` into C++ and it's 300× faster. ("Ought to be free" because it's _O(1)_.)

On to step 2, the remaining 3/4 of the bottleneck...

**Comment by jpivarski** [comment]
Okay, it was actually two more steps (three optimizations total), but now the original code in https://github.com/scikit-hep/awkward-1.0/issues/538#issue-742805318 goes from (without fixes):

```
% python testy3.py 
number of variants 50
time to make variants: 0.3175015449523926
time to assign with __setitem__: 5.913200616836548
time to assign with rezip: 0.21291756629943848
% python testy3.py 
number of variants 50
time to make variants: 0.32436084747314453
time to assign with __setitem__: 6.110222101211548
time to assign with rezip: 0.2174851894378662
% python testy3.py 
number of variants 50
time to make variants: 0.31608152389526367
time to assign with __setitem__: 5.901622295379639
time to assign with rezip: 0.21786880493164062
```

to (with fixes):

```
% python testy3.py 
number of variants 50
time to make variants: 0.08354496955871582
time to assign with __setitem__: 0.11821508407592773
time to assign with rezip: 0.028712034225463867
% python testy3.py 
number of variants 50
time to make variants: 0.08423042297363281
time to assign with __setitem__: 0.11730170249938965
time to assign with rezip: 0.02914595603942871
% python testy3.py 
number of variants 50
time to make variants: 0.08287811279296875
time to assign with __setitem__: 0.11925029754638672
time to assign with rezip: 0.028862953186035156
```

As a bottom line, the `__setitem__` method (which was the main target) is now **50× faster**. (Speedups for the individual parts were much larger, but that made them not bottlenecks anymore. Beyond the three fixes in #540, they're not single-item hotspots.)
